board/safety/safety_volkswagen.h

```diff
@@ -0,0 +1,169 @@
+const int VW_MAX_STEER = 300;               // 3.0 nm
+const int VW_MAX_RT_DELTA = 130;            // max delta torque allowed for real time checks
+const uint32_t VW_RT_INTERVAL = 250000;     // 250ms between real time checks
+const int VW_MAX_RATE_UP = 10;
+const int VW_MAX_RATE_DOWN = 10;
+const int VW_DRIVER_TORQUE_ALLOWANCE = 100;
+const int VW_DRIVER_TORQUE_FACTOR = 4;
+
+int vw_ignition_started = 0;
+struct sample_t vw_torque_driver;           // last few driver torques measured
+int vw_rt_torque_last = 0;
+int vw_desired_torque_last = 0;
+uint32_t vw_ts_last = 0;
+
+// Safety-relevant CAN messages for the Volkswagen MQB platform.
+#define MSG_EPS_01              0x09F
+#define MSG_ACC_06              0x122
+#define MSG_HCA_01              0x126
+#define MSG_GRA_ACC_01          0x12B
+#define MSG_LDW_02              0x397
+#define MSG_KLEMMEN_STATUS_01   0x3C0
+
+static void volkswagen_init(int16_t param) {
+  UNUSED(param); // May use param in the future to indicate MQB vs PQ35/PQ46/NMS vs MLB, or wiring configuration.
+  controls_allowed = 0;
+  vw_ignition_started = 0;
+}
+
+static int volkswagen_ign_hook(void) {
+  // While we do monitor VW Terminal 15 (ignition-on) state, we are not currently acting on it. We may do so in the
+  // future for harness integrations at the camera (where we only have T30 unswitched power) instead of the gateway
+  // (where we have both T30 and T15 ignition-switched power). For now, use the default GPIO pin behavior.
+
+  // return vw_ignition_started;
+  return -1;
+}
+
+static void volkswagen_rx_hook(CAN_FIFOMailBox_TypeDef *to_push) {
+  int bus = GET_BUS(to_push);
+  int addr = GET_ADDR(to_push);
+
+  // Monitor Klemmen_Status_01.ZAS_Kl_15 for Terminal 15 (ignition-on) status, but we make no use of it at the moment.
+  if ((bus == 0) && (addr == MSG_KLEMMEN_STATUS_01)) {
+    vw_ignition_started = (GET_BYTE(to_push, 2) & 0x2) >> 1;
+  }
+
+  // Update driver input torque samples from EPS_01.Driver_Strain for absolute torque, and EPS_01.Driver_Strain_VZ
+  // for the direction.
+  if ((bus == 0) && (addr == MSG_EPS_01)) {
+    int torque_driver_new = GET_BYTE(to_push, 5) | ((GET_BYTE(to_push, 6) & 0x1F) << 8);
+    int sign = (GET_BYTE(to_push, 6) & 0x80) >> 7;
+    if (sign == 1) {
```

Comment:
so weird they did not make it a 14 bits signed signal. the 2 bits gap between the message and the sign is very unusual.

Reply:
Germans, what are you gonna do. Anyway, they are somewhat consistent about putting the sign of directional values in the high-bit of the MSB byte. Examples include the driver input torque and steering angle in EPS_01 and the heading control assist torque in HCA_01.